File: webapp/src/stores/immich.ts
import { defineStore } from 'pinia';
import { api } from '../api';
import { getApiError } from '../utils/errors';

interface CacheStatus {
  enabled: boolean;
  count: number;
  size_bytes: number;
  size_human: string;
  max_images: number;
  max_size_mb: number;
  populating: boolean;
}

export const useImmichStore = defineStore('immich', {
  state: () => ({
    count: 0,
    albums: [] as any[],
    syncedAlbums: [] as any[],
    loading: false,
    error: null as string | null,
    cacheStatus: null as CacheStatus | null,
  }),
  actions: {
    async fetchCount() {
      try {
        const res = await api.get('/immich/count');
        this.count = res.data.count || 0;
      } catch (e) {
        console.error('Failed to fetch Immich photo count', e);
      }
    },

    async fetchAlbums() {
      this.loading = true;
      this.error = null;
      try {
        const res = await api.get('/immich/albums');
        this.albums = res.data;
      } catch (e) {
        this.error = getApiError(e);
        throw e;
      } finally {
        this.loading = false;
      }
    },

    async fetchSyncedAlbums() {
      try {
        const res = await api.get('/albums?source=immich');
        this.syncedAlbums = res.data || [];
        return this.syncedAlbums;
      } catch (e) {
        console.error('Failed to fetch synced Immich albums', e);
        throw e;
      }
    },

    async saveSyncAlbums(payload: {
      album_ids: string[];
      favorites: boolean;
      all: boolean;
      memories: boolean;
    }) {
      this.loading = true;
      try {
        await api.post('/immich/sync-albums', payload);
        await this.fetchSyncedAlbums();
        await this.fetchCount();
      } finally {
        this.loading = false;
      }
    },

    async testConnection() {
      this.loading = true;
      try {
        const res = await api.post('/immich/test');
        await this.fetchCount();
        return res.data;
      } finally {
        this.loading = false;
      }
    },

    async sync() {
      this.loading = true;
      try {
        await api.post('/immich/sync');
        await this.fetchCount();
      } finally {
        this.loading = false;
      }
    },

    async fetchCacheStatus() {
      try {
        const res = await api.get('/immich/cache/status');
        this.cacheStatus = res.data;
      } catch (e) {
        console.error('Failed to fetch Immich cache status', e);
      }
    },

    async populateCache() {
      try {
        await api.post('/immich/cache/populate');
        await this.fetchCacheStatus();
      } catch (e) {
        this.error = getApiError(e);
        throw e;
      }
    },

    async clearCache() {
      try {
        await api.post('/immich/cache/clear');
        await this.fetchCacheStatus();
      } catch (e) {
        this.error = getApiError(e);
        throw e;
      }
    },
  },
});

File: backend/internal/handler/immich.go
package handler

import (
	"net/http"

	"github.com/aitjcize/esp32-photoframe-server/backend/internal/service"
	"github.com/labstack/echo/v4"
)

type ImmichHandler struct {
	immich *service.ImmichService
}

func NewImmichHandler(s *service.ImmichService) *ImmichHandler {
	return &ImmichHandler{immich: s}
}

func (h *ImmichHandler) TestConnection(c echo.Context) error {
	if err := h.immich.TestConnection(); err != nil {
		return respondError(c, http.StatusBadRequest, err.Error())
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "ok"})
}

func (h *ImmichHandler) ListAlbums(c echo.Context) error {
	albums, err := h.immich.ListAlbums()
	if err != nil {
		return respondError(c, http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, albums)
}

// SetSyncAlbums defines which Immich albums (real + virtual modes) to sync.
// POST /api/immich/sync-albums
func (h *ImmichHandler) SetSyncAlbums(c echo.Context) error {
	var req struct {
		AlbumIDs  []string `json:"album_ids"`
		Favorites bool     `json:"favorites"`
		All       bool     `json:"all"`
		Memories  bool     `json:"memories"`
	}
	if err := c.Bind(&req); err != nil {
		return respondError(c, http.StatusBadRequest, "invalid request")
	}
	if err := h.immich.SetSyncAlbums(req.AlbumIDs, req.Favorites, req.All, req.Memories); err != nil {
		return respondError(c, http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "ok"})
}

// CacheStatus returns the current state of the Immich image cache.
// GET /api/immich/cache/status
func (h *ImmichHandler) CacheStatus(c echo.Context) error {
	cache := h.immich.GetCacheService()
	if cache == nil {
		return c.JSON(http.StatusOK, service.CacheStatus{Enabled: false})
	}
	return c.JSON(http.StatusOK, cache.Status())
}

// CachePopulate triggers an immediate background cache population cycle.
// POST /api/immich/cache/populate
func (h *ImmichHandler) CachePopulate(c echo.Context) error {
	cache := h.immich.GetCacheService()
	if cache == nil {
		return respondError(c, http.StatusBadRequest, "cache service not available")
	}
	if !cache.Enabled() {
		return respondError(c, http.StatusBadRequest, "immich cache is not enabled")
	}
	cache.PopulateNow()
	return c.JSON(http.StatusOK, map[string]string{"status": "population started"})
}

// CacheClear deletes all cached images from disk and the database.
// POST /api/immich/cache/clear
func (h *ImmichHandler) CacheClear(c echo.Context) error {
	cache := h.immich.GetCacheService()
	if cache == nil {
		return respondError(c, http.StatusBadRequest, "cache service not available")
	}
	if err := cache.ClearCache(); err != nil {
		return respondError(c, http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "cache cleared"})
}

// Sync / SyncStatus / Clear / Count are served by the generic PhotoSyncHandler
// (see photosync_handler.go) — identical across photo sources.
